library/psa_crypto_mac.c

```diff
@@ -383,23 +392,39 @@ psa_status_t mbedtls_psa_mac_update(
 
 static psa_status_t psa_mac_finish_internal(
     mbedtls_psa_mac_operation_t *operation,
-    uint8_t *mac, size_t mac_size)
+    uint8_t *mac, size_t mac_size, size_t *mac_length)
 {
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    size_t actual_mac_length = PSA_MAC_TRUNCATED_LENGTH(operation->alg);
+    /* We know that operation->alg is truncated after the mac setup function has
+     * been called. Hence, if the PSA_MAC_TRUNCATED_LENGTH returns 0  this means
+     * that the mac length should be 64 as specified in the definition of
+     * PSA_ALG_MAC_TRUNCATION_MASK
```

Comment:
No:

> the value 0 means the untruncated algorithm, whatever its length is

There's a macro `PSA_MAC_LENGTH` for that. See `psa_mac_finalize_alg_and_key_validation`, which is what is used to set `operation->mac_size`. Why not use `operation->mac_size` here anyway?

Reply:
Perhaps there is a straight forward way that I'm missing, but here are my comments and proposals nevertheless:

1. From what I understand, `operation->mac_size` is for `psa_mac_operation_t` but from this function we are dealing with `mbedtls_psa_mac_operation_t`, that I don't think has this field?

2. `PSA_MAC_LENGTH` needs key type and key bit len information, so the alternative would look something like this:
```

  +    size_t actual_mac_length = mac_size;
   
  -    size_t actual_mac_length = PSA_MAC_TRUNCATED_LENGTH(operation->alg);
  -    /* We know that operation->alg is truncated after the mac setup function has
  -     * been called. Hence, if the PSA_MAC_TRUNCATED_LENGTH returns 0  this means
  -     * that the mac length should be 64 as specified in the definition of
  -     * PSA_ALG_MAC_TRUNCATION_MASK
  -     */
  -    if (actual_mac_length == 0) {
  -        actual_mac_length = 64;
  -    }
  +#if defined(MBEDTLS_PSA_BUILTIN_ALG_CMAC)
  +    actual_mac_length = PSA_MAC_LENGTH(PSA_KEY_TYPE_AES,
  +                                       operation->ctx.cmac.key_bitlen,
  +                                       operation->alg);
  +#elif defined(MBEDTLS_PSA_BUILTIN_ALG_HMAC)
  +    actual_mac_length = PSA_MAC_LENGTH(PSA_KEY_TYPE_HMAC, 0, operation->alg);
  +#endif

```  
Is this what you had in mind? I find it a bit uncomfortable to use `PSA_MAC_LENGTH(PSA_KEY_TYPE_HMAC, 0, operation->alg);`

Reply:
Ah, the core/driver separation strikes again. Indeed the driver doesn't have access to `operation->mac_size`.

During `finish`, we don't have easy access to the key type. So `PSA_MAC_LENGTH` is indeed not a good solution.

We can store the full length and the truncated length in the driver operation (`mbedtls_psa_mac_operation_t`) in the setup entry point (`psa_mac_setup`). That would cost a few bytes of code size in `setup`, but it would save at least that cost in `finish`.

For MAC accelerator drivers, arguably, it would make sense to just make the driver calculate the whole-length MAC, and have the core truncate. But that would create a difference with secure element drivers. They have to calculate the truncated MAC, since they might have a policy that prevents getting the full MAC. I'm not sure if simplifying accelerators is worth complicating the driver interface in this way.

Reply:
I hope I understood correctly and did all the required changes now :)